source/agora/node/Ledger.d

```diff
@@ -125,37 +125,72 @@ public class Ledger
             throw new Exception("Genesis block loaded from disk is " ~
                 "different from the one in the config file");
 
-        // need to regenerate the UTXO set, starting from the genesis block
+        // need to rebuild the UTXO set and Validator set,
+        // starting from the genesis block
         if (this.utxo_set.length == 0)
         {
             Block block;
+            Block last_read_block = gen_block;
+
+            ManagedDatabase.beginBatch();
+            scope (failure) ManagedDatabase.rollback();
+
+            // clear validator set
+            this.enroll_man.removeAllValidators();
+
             foreach (height; 0 .. this.last_block.header.height + 1)
             {
                 this.storage.readBlock(block, Height(height));
+                if (height == 0)
+                {
+                    if (auto reason = block.isGenesisBlockInvalidReason())
+                        throw new Exception(
+                            "Genesis block loaded from disk is invalid: " ~
+                            reason);
+                }
+                else
+                {
+                    size_t active_enrollments = enroll_man.getValidatorCount(
```

Comment:
```suggestion
                    const active_enrollments = enroll_man.getValidatorCount(
```

Prefer using `const` or `const size_t`, either works. If we don't modify variables it's always best to make the const.